tests/e2e/e2e_pod_client_server_test.go

```diff
@@ -136,14 +140,133 @@ func testTraffic(withSourceKubernetesService bool) {
+
+		It("Tests TCP traffic for client pod -> server pod", func() {
+			installOpts := Td.GetOSMInstallOpts()
+			// Install OSM
+			Expect(Td.InstallOSM(installOpts)).To(Succeed())
+
+			// Load TCP server image
+			Expect(Td.LoadImagesToKind([]string{"tcp-echo-server"})).To(Succeed())
+
+			// Create Test NS
+			for _, n := range ns {
+				Expect(Td.CreateNs(n, nil)).To(Succeed())
+				Expect(Td.AddNsToMesh(true, n)).To(Succeed())
+			}
+
+			destinationPort := 80
+
+			// Get simple pod definitions for the TCP server
+			svcAccDef, podDef, svcDef := Td.SimplePodApp(
+				SimplePodAppDef{
+					Name:        destName,
+					Namespace:   destName,
+					Image:       fmt.Sprintf("%s/tcp-echo-server:%s", installOpts.ContainerRegistryLoc, installOpts.OsmImagetag), // FIXME
+					Command:     []string{"/tcp-echo-server"},
+					Args:        []string{"--port", fmt.Sprintf("%d", destinationPort)},
+					Ports:       []int{destinationPort},
+					AppProtocol: AppProtocolTCP,
+				})
+
+			_, err := Td.CreateServiceAccount(destName, &svcAccDef)
+			Expect(err).NotTo(HaveOccurred())
+			_, err = Td.CreatePod(destName, podDef)
+			Expect(err).NotTo(HaveOccurred())
+			dstSvc, err := Td.CreateService(destName, svcDef)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Expect it to be up and running in it's receiver namespace
+			Expect(Td.WaitForPodsRunningReady(destName, 120*time.Second, 1)).To(Succeed())
+
+			srcPod := setupSource(sourceName, withSourceKubernetesService)
+
+			By("Creating SMI policies")
+			// Deploy allow rule client->server
+			trafficTarget := Td.CreateSimpleTCPAllowPolicy(
+				SimpleAllowPolicy{
+					RouteGroupName:    "routes",
+					TrafficTargetName: "test-target",
+
+					SourceNamespace:      sourceName,
+					SourceSVCAccountName: sourceName,
+
+					DestinationNamespace:      destName,
+					DestinationSvcAccountName: destName,
+				})
+
+			// Configs have to be put into a monitored NS
+			_, err = Td.CreateTrafficTarget(sourceName, trafficTarget)
+			Expect(err).NotTo(HaveOccurred())
+
+			// All ready. Expect client to reach server
+			requestMsg := "test request"
+			clientToServer := TCPRequestDef{
+				SourceNs:        sourceName,
+				SourcePod:       srcPod.Name,
+				SourceContainer: sourceName,
+
+				DestinationHost: fmt.Sprintf("%s.%s", dstSvc.Name, dstSvc.Namespace),
+				DestinationPort: destinationPort,
+				Message:         requestMsg,
+			}
+
+			srcToDestStr := fmt.Sprintf("%s -> %s:%d",
+				fmt.Sprintf("%s/%s", sourceName, srcPod.Name),
+				clientToServer.DestinationHost, clientToServer.DestinationPort)
+
+			cond := Td.WaitForRepeatedSuccess(func() bool {
+				result := Td.TCPRequest(clientToServer)
+
+				if result.Err != nil {
+					Td.T.Logf("> (%s) TCP Req failed, response: %s, err: %s",
+						srcToDestStr, result.Response, result.Err)
+					return false
+				}
+
+				// Ensure the echo response contains request message
+				if !strings.Contains(result.Response, requestMsg) {
```

Comment:
nit: should we consider exact equal since we expect echo? (is nc returning a clean echo or do you have some other stdout text hanging around?)

Reply:
nope, server prefixes the response with a prefix. So if client sends `foo`, server responds `echo response: foo`